src/main/java/org/apache/commons/imaging/common/ImageBuilder.java

```diff
@@ -68,6 +69,26 @@
      * requirements for the ImageBuilder or resulting BufferedImage.
      */
     public ImageBuilder(final int width, final int height, final boolean hasAlpha) {
+        this(true, width, height, hasAlpha);
+    }
+    
+    /**
+     * Construct an ImageBuilder instance.
+     * <p>
+     * Usually image data is stored bottom-up, that means that the last row in the
+     * image data is the first row of pixels in the image as displayed.
+     * Sometimes image data are stored top-down: The first row in the image data
+     * is also the first row of pixels in the image as displayed.
+     * With this constructor you can define the direction.
+     * @param readBottomUp {@code true}: the image data are stored bottom-up,
+     * {@code false} the image data are stored top-down
+     * @param width the width of the image to be built
+     * @param height the height of the image to be built
+     * @param hasAlpha indicates whether the image has an alpha channel
+     * (the selection of alpha channel does not change the memory
+     * requirements for the ImageBuilder or resulting BufferedImage.
+     */
+    public ImageBuilder(final boolean readBottomUp, final int width, final int height, final boolean hasAlpha) {
```

Comment:
On 23.02.2015 20:43, Benedikt Ritter
      wrote:
    
    
      In src/main/java/org/apache/commons/imaging/common/ImageBuilder.java:
      > +     * Construct an ImageBuilder instance.
> +     * <p>
> +     * Usually image data is stored bottom-up, that means that the last row in the
> +     * image data is the first row of pixels in the image as displayed.
> +     * Sometimes image data are stored top-down: The first row in the image data
> +     * is also the first row of pixels in the image as displayed.
> +     * With this constructor you can define the direction.
> +     * @param readBottomUp {@code true}: the image data are stored bottom-up,
> +     * {@code false} the image data are stored top-down
> +     * @param width the width of the image to be built
> +     * @param height the height of the image to be built
> +     * @param hasAlpha indicates whether the image has an alpha channel
> +     * (the selection of alpha channel does not change the memory
> +     * requirements for the ImageBuilder or resulting BufferedImage.
> +     */
> +    public ImageBuilder(final boolean readBottomUp, final int width, final int height, final boolean hasAlpha) {

      Is there a way for us to detect whether the image is bottom up?
      —
        Reply to this email directly or view
          it on GitHub.
      
        
          
          
        
        
      
    
    > Is there a way for us to detect whether the image is bottom up?
    
    I found a much simpler way without modifying ImageBuilder. Now it is
    only a matter inside org.apache.commons.imaging.formats.bmp package.
    For bmp image files: If the header has a positive height, it is
    bottom-up. You may read BitmapHeaderInfo.isBottomUpBitmap: If it is
    true the the height was positive and the bitmap is to read
    bottom-up, if false the height was negative and it is to read
    top-down.
    
    Before the patch the BitmapHeaderInfo.height may be negative, now it
    is always positive.
    
    I read the specs and docs again: Miroslav Gloub is right: If the
    image data are compressed, the height must be positive.
    -- 
Michael Groß